keps/sig-api-machinery/5366-graceful-leader-transition/README.md

```diff
@@ -0,0 +1,940 @@
+nitty-gritty.
+-->
+
+The main control loops for `kube-controller-manager`, `kube-scheduler`, and
+`cloud-controller-manager` will be updated to support graceful leader
+transitions. When a leader fails to renew its lease, instead of exiting, the
+component will rely on client-go's leader election mechanism to cancel the
+context, and stop its internal controllers. It will then immediately return to a
+follower state where it will attempt to reacquire the lease.
+
+This change will be guarded by a new feature gate, `GracefulLeaderTransition`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+In an HA configuration, cloud provider A wants to balance controllers over
+multiple control plane instances. With graceful transitions, multiple locks can
+be used by a single KCM instance such that a subset of components run under each
+lock.
+
+#### Story 2
+
+A extension developer has controller manager that should dynamically start and
+shutdown controllers based on cluster state (such as the registration of
+resources that declare how a CRD should be reconciled). The extension developer
+requires that controllers shutdown gracefully so that only the controller loops
+that SHOULD be running continue to run, and that no resources are leaked over
+time as controllers are started and stopped.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Kubernetes leader elected controllers and the scheduler have been running
+without graceful shutdown for years. 
+
+Risk 1: Resource exhaustion: Memory leaks may exist in the processes that were
+previously masked by doing a full shutdown and restart loop. 
+
+- Severity: Medium high
+- Controllers will continue to function (potentially in degraded state due to
+  lack of resources), and may be restarted frequently. However, cluster should
+  continue to function.
+
+Risk 2: Wedged KCM: There is a risk that controllers and the scheduler are not
+properly respecting context shutdowns. This can either result in multiple
+instances of controllers running or no instances running despite the lock being
+held.
+
+- Severity: High
+- Breaking mutual exclusion guarantees can put the cluster into a non-desirable
+  state. A manual user intervention is possible but if the problem is triggered
+  due to a problematic component, the issue will resurface and the best path for
+  mitigation is to turn off the feature.
```

Comment:
I'd like some kind of detection and test built for this.

Reply:
This is in the test that only ONE control loop should be running for proper mutual exclusion.

Reply:
The runtime detection below for alpha satisfies.